src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -642,5 +711,212 @@ private static void ThrowMarkedCtorDoesNotTakeAllProvidedArguments()
+                        ? arguments![parameter1.ArgumentIndex]
+                        : GetService(
+                            serviceProvider,
+                            parameter1.ParameterType,
+                            declaringType,
+                            parameter1.HasDefaultValue)) ?? parameter1.DefaultValue;
+                    break;
+            }
+
+            return invoker.Invoke(arg1, arg2, arg3, arg4);
+        }
+
+        private static object ReflectionFactoryCanonicalSpan(
+            ConstructorInvoker invoker,
+            FactoryParameterContext[] parameters,
+            Type declaringType,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
+
+            object?[] constructorArguments = new object?[parameters.Length];
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                ref FactoryParameterContext parameter = ref parameters[i];
+                constructorArguments[i] = ((parameter.ArgumentIndex != -1)
+                    // Throws a NullReferenceException if arguments is null. Consistent with expression-based factory.
+                    ? arguments![parameter.ArgumentIndex]
+                    : GetService(
+                        serviceProvider,
+                        parameter.ParameterType,
+                        declaringType,
+                        parameter.HasDefaultValue)) ?? parameter.DefaultValue;
+            }
+
+            return invoker.Invoke(constructorArguments.AsSpan());
+        }
+
+        private static object ReflectionFactoryDirect(
+            ConstructorInvoker invoker,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
+
+            if (arguments is null)
+                ThrowHelperNullReferenceException(); //AsSpan() will not throw NullReferenceException.
+
+            return invoker.Invoke(arguments.AsSpan());
+        }
+
+        /// <summary>
+        /// For consistency with the expression-based factory, throw NullReferenceException.
+        /// </summary>
+        [DoesNotReturn]
+        private static void ThrowHelperNullReferenceException()
+        {
+            throw new NullReferenceException();
+        }
+#elif NETSTANDARD2_1_OR_GREATER || NETCOREAPP
+        private static object ReflectionFactoryCanonical(
+            ConstructorInfo constructor,
+            FactoryParameterContext[] parameters,
+            Type declaringType,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
+
+            object?[] constructorArguments = new object?[parameters.Length];
```

Comment:
Not saying we should *but* we could pool this array now that we have Span support.